File: 32. Longest Valid Parentheses.py
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        '''
        Using a stack to keep track the valid parentheses,
        when it comes to '(', we push it into the stack,
        when it comes to ')', we pop from the stack and we check if 
        the stack is empty, if empty, we put the index of this element into the stack
        and we calculate the length of the valid parentheses
        '''
        stack = [-1]
        maxlen = 0
        for i, char in enumerate(s):
            if char == '(':
                stack.append(i)
            else:
                stack.pop(-1)
                if not stack:
                    stack.append(i)
                maxlen = max(maxlen, i-stack[-1])
        return maxlen

class Solution:
    def longestValidParentheses(self, s: str) -> int:
        '''
        https://leetcode.com/articles/longest-valid-parentheses/#approach-2-using-dynamic-programming
        cannot understand approach 4, try using stack is easiest to understand
        '''
        maxlen = 0
        dp = [0 for _ in s]
        for i in range(1, len(s)):
            if s[i] == ')':
                if s[i-1] == '(':
                    dp[i] = dp[i-2]+2
                elif i - dp[i-1] > 0 and s[i-dp[i-1]-1] == '(':
                    dp[i] = dp[i-1] +  dp[i - dp[i - 1] - 2] + 2
                maxlen = max(maxlen, dp[i])
        return maxlen
                    

File: 32. Longest Valid Parentheses.java
/**
 * stack one pass solution,
 * use stack to keep track of valid parentheses
 */
class Solution {
    public int longestValidParentheses(String s) {
        Deque<Integer> stack = new ArrayDeque<>();
        stack.push(-1);
        int maxlen = 0;
        for (int i=0; i<s.length(); i++) {
            if (s.charAt(i) == '(') {
                stack.push(i);
            } else {
                stack.pop();
                if (stack.isEmpty()) {
                    stack.push(i);
                }
                maxlen = Math.max(maxlen, i-stack.peek());
            }
        }
        return maxlen;
    }
}

/**
 * without stack solution,
 * two pass,
 * 1. from left to right:
 * if we find left == right we record it,
 * if we find right is more, we need to discard previous
 * 2. from right to left:
 * if we find left == right we record it,
 * if we find left is more, we need to discard previous
 */
class Solution {
    public int longestValidParentheses(String s) {
        int left=0, right=0, maxlen=0;
        for (char c : s.toCharArray()) {
            if (c == '(') {
                left++;
            }
            if (c == ')') {
                right++;
            }
            if (left == right) {
                maxlen = Math.max(maxlen, 2*right);
            }
            if (right > left) {
                right = 0;
                left = 0;
            }
        }
        left = 0;
        right = 0;
        for (int i=s.length()-1; i>=0; i--) {
            char c = s.charAt(i);
            if (c == '(') {
                left++;
            }
            if (c == ')') {
                right++;
            }
            if (left == right) {
                maxlen = Math.max(maxlen, 2*right);
            }
            if (left > right) {
                right = 0;
                left = 0;
            }
        }
        return maxlen;
    }
}